Modify rule S6320: Adapt to LaYC format (#3158)
SonarSource/rspec: rules/S6320/java/rule.adoc

> == Why is this an issue?

Reviewer: For level 1 LaYC format, there needs to be an introductory section without a heading before the `Why is this an issue?` section. This introduction section is also displayed separately in SonarQube.

A cast operation allows an object to be converted from one type to another.

> The compiler raises an error if it can determine that the target type is incompatible with the declared type of the object, otherwise it accepts the cast.

Reviewer: Nitpick:
```suggestion
The compiler raises an error if it can determine that the declared target type is incompatible with the actual runtime type of the object, otherwise it accepts the cast.
```

However, depending on the actual runtime type of the object, a cast operation may fail at runtime.
When a cast operation fails, a `ClassCastException` is thrown.

> This type of exception is usually unexpected, so it is not caught and the application crashes.

Reviewer: Regarding a `What is the Potential Impact?` section for this rule:

The impact of rules about exceptions being raised, is usually the same, see for example [the potential impact section for S6466](https://github.com/SonarSource/rspec/blob/fa6a5beb638fe7e4502104cdd899ae2e083bcea4/rules/S6466/impact.adoc).

Apparently there is a `shared_content` folder for sections that are shared across rules, see [here](https://github.com/SonarSource/rspec/blob/fa6a5beb638fe7e4502104cdd899ae2e083bcea4/docs/description.adoc#share-content-between-rules).
Maybe we should move the impact description of S6466 there and also import it in this rule?

> 

Reviewer: Regarding a `What is the Potential Impact?` section for this rule:

The impact of rules about exceptions being raised, is usually the same, [see the potential impact section for S6466](https://github.com/SonarSource/rspec/blob/fa6a5beb638fe7e4502104cdd899ae2e083bcea4/rules/S6466/impact.adoc).

Apparently there is a `shared_content` folder for sections that are shared across rules, see [here](https://github.com/SonarSource/rspec/blob/fa6a5beb638fe7e4502104cdd899ae2e083bcea4/docs/description.adoc#share-content-between-rules).
Maybe we should move the impact description of S6466 there and also import it here?

> == How can I fix it?

Reviewer: This section has to be named exactly `How to fix it`, otherwise no tab will be created for it in the SonarQube UI

When an object is cast, the code makes assumptions about the type of the object.
A `ClassCastException` indicates that this assumption has been broken.
If the assumption is reasonable, then some prior logic should ensure that only objects of a compatible type can be cast.
You should try to identify the code responsible for these checks and fix it.

=== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----
String hexString(Object o) {
  return Integer.toHexString((Integer) o); // Noncompliant if hexString is called with a String for example 
}
----

=== Compliant solution

One possible solution is to change `hexString` to only accept integers and adapt call sites.

[source,java,diff-id=1,diff-type=compliant]
----
String hexString(Integer i) {
  return Integer.toHexString(i);
}
----

Another one is to return a default value for types that are not `Integer`.

[source,java,diff-id=1,diff-type=compliant]
----
String hexString(Object o) {
  if (o instanceof Integer) {
    return Integer.toHexString((Integer) o);
  }
  return "0x0";
}
----

It depends on which part of the code should be responsible to handle non-Integer types.

> 

Reviewer: This is not something we need to worry about right now, since we are not aiming for level 2, but I still wanted to leave a comment about the `How does it work?` section that is technically part of `How to fix it`.
This comment is not really a part of the review, but maybe something to discuss in our feedback meeting with John/Loris.

We are given an example of a `How to fix it` section, and a `How does it work?` section [here](https://next.sonarqube.com/sonarqube/coding_rules?languages=java&q=xss&open=javasecurity%3AS5131).
However, the difference between the two really is not clear to me.

The [LaYC specification](https://xtranet-sonarsource.atlassian.net/wiki/spaces/PM/pages/2715320326/LaYC+Rule+Content+Specification#How-can-I-fix-it?) gives me the impression that examples go directly into the `How to fix it` section, while general advise and best practices should go into `How does it work?`.
And that is also roughly what we see in the XSS example linked above.

Author: I added this here because I don't think that giving a simple example and the easiest solution helps the user. I think it's good if they think about where the logic should change (i.e. which part of the code is responsible for it)

I will skip the "How to fix it" and "Pitfalls" subsections for now, because I think it will just repeat information. We can make feedback on this for sure.

=== Going the extra mile

Casting is considered an anti-pattern in object-oriented programming.
It is sometimes necessary, but there often is a better alternative.
Consider:

* Polymorphism
* The visitor design pattern
* Pattern matching

> They come with some garantees from the compiler, which makes your code more reliable.

Reviewer: ```suggestion
They come with some guarantees from the compiler, which make your code more reliable.
```

==== Noncompliant code example

[source,java,diff-id=2,diff-type=noncompliant]
----
int foo(Shape shape) {
  if (shape instanceof Circle) {
    Circle circle = (Circle) shape;
    // Code for objects of type Circle
  } else if (shape instanceof Square) {
    Square square = (Square) shape;
    // Code for objects of type Square
  }
  // Default
}
----

==== Compliant solution

[source,java,diff-id=2,diff-type=compliant]
----
int foo(Shape shape) {

>   return shape.fooValue(...); // Code was moved in subclasses

Reviewer: Nitpick
```suggestion
  return shape.fooValue(...); // Code was moved into subclasses

}
----

or

[source,java,diff-id=2,diff-type=compliant]
----
int foo(Shape shape) {

>   return shape.accept(new FooVisitor()); // Code was moved in FooVisitor

Reviewer: Nitpick:
```suggestion
  return shape.accept(new FooVisitor()); // Code was moved into FooVisitor
```

}
----

or

[source,java,diff-id=2,diff-type=compliant]
----
// Java 14+ required
int foo(Shape shape) {
  if (shape instanceof Circle c) {
    // Code for objects of type Circle
  } else if (shape instanceof Square s) {
    // Code for objects of type Square
  }
}
----

> == More info

Reviewer: Unfortunately, the LaYC spec is again not very accurate when it comes to the heading names to use in practice.
To have this section displayed in the `More Info` tab in SonarQube, it apparently needs to be named "Resources" :/

```suggestion
== Resources
```

> 

Reviewer: Nitpick: I am not entirely sure about this, but I believe the links in "More Info" need to be organized into category headings. I.e. "Documentation" or "Articles & Blog Posts"

```suggestion

=== Documentation

* https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ClassCastException.html[ClassCastException]

=== Articles & blog posts

```

* https://refactoring.guru/design-patterns/visitor[Refactoring Guru: Visitor]

> * https://openjdk.org/jeps/394[JEP 394: Pattern Matching for instanceof]

Reviewer: ```suggestion
* https://openjdk.org/jeps/394[JEP 394: Pattern Matching for instanceof]

=== Standards

* https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.16
* https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2[The JLS on instanceof]
```
